### ISSUE TITLE
`drmad/cpu_ver/experiments/exp_learning_rate` issue

### ISSUE THREAD

**Issue by bkj** [bigaidream-projects/drmad#10]
When I run `mad.py` and `exact.py` in `drmad/cpu_ver/experiments/exp_learning_rate` I get totally different results.

MAD:
![screen shot 2017-08-31 at 7 34 12 pm](https://user-images.githubusercontent.com/6086781/29949554-82705332-8e83-11e7-8700-355793a148c4.png)

Exact:
![screen shot 2017-08-31 at 7 34 15 pm](https://user-images.githubusercontent.com/6086781/29949551-7f59e4d8-8e83-11e7-94bf-a17d2253f10e.png)

Can you verify that this is the expected behavior?

Thanks
Ben

**Comment by bigaidream** [comment]
Compared to static hyperparameters (e.g. L2 regularizaitons), it seems that dynamic hyperparameters (e.g. learning rates) are more sensitive to super-parameters like hyper-learning rates. Actually, the experiments with learning rates and momentum are not included in the original paper. 

---

After that paper, we tried some super-parameters, but the ones in the repo might be not the best ones. 

Thanks,
Jie

**Comment by bkj** [comment]
Do you have any suggestions about better super parameters?  Would be very interesting to get this working

**Comment by bigaidream** [comment]
I did have some decent results, but didn't save them as my previous focus was on static hyperparameters. I will be running some experiments on my servers, and tell you later.

**Comment by bigaidream** [comment]
One particular setting of super-parameters is

```python
N_iters = 500
meta_alpha = 0.06
N_meta_iter = 10
```

According to the loss values, in this setting `mad.py` seems to be similar to `exact.py`. 

---

To be honest, as for the state-of-the-art research on tuning learning rates, I'd suggest taking a look at [YellowFin](https://github.com/JianGoForIt/YellowFin_Pytorch) and [SGD with restart](https://github.com/loshchil/SGDR).

**Comment by bkj** [comment]
Fantastic -- thanks! Will test this later and let you know how it goes.

**Comment by bigaidream** [comment]
The whole purpose of DrMAD method is to show that BP for hyperparameters (especially static) can be done in an approximate way for simple deep models on toy datasets. It's not a practical way of tuning large-scale deep models yet, especially for dynamic hyperparameters on realistic datasets. It implicitly assumes that the hyperparameter error surface is convex. Furthermore, it has to deal with hyper-gradient vanishing problems, which I did not solve.

**Comment by michaelstjules** [comment]
I think one of the main reason you wouldn't want to tune per-iteration learning rates with DrMAD is because the hypergradients with respect to these learning rates are probably sensitive to the SGD paths actually taken, and DrMAD only very crudely approximates these paths. L2 penalties hyperparameters, on the other hand, are shared between all of the SGD iterates, so the hypergradients with respect to them may be more easily approximated with DrMAD. Learning rates shared between iterations might be more feasible, but I'm not sure if anyone's looked into this. 

I played around some more with DrMAD and other gradient-based hyperparameter optimization methods here, in case you're interested: https://uwaterloo.ca/computational-mathematics/sites/ca.computational-mathematics/files/uploads/files/michael_st._jules.pdf

I think it's worth pointing out here that with DrMAD and its straight uniformly-spaced path between initial and final parameters, the number of reverse iterations (to compute the DrMAD hypergradient) should not continue to grow with the number of forward iterations (during SGD). E.g. 200 reverse iterations was "safe" for 200 to 200,000 forward iterations, but 2000 or more reverse iterations leads to exploding hypergradients (for L2 penalty hyperparameters), for 200 to 200,000 forward iterations. 

I'll try to upload the code soon.

**Comment by bkj** [comment]
Interesting -- I'll take a look at the paper.  Would love to see code. 

Makes a lot of sense that the per-iteration learning rates shouldn't really work with an approximate path.

**Comment by michaelstjules** [comment]
I've just uploaded the code: https://github.com/michaelstjules/hypergrad-experiments

**Comment by bigaidream** [comment]
Interesting work. It seems that for bigger problems, the straight uniformly-spaced path will lead to exploding hypergradients problems. More sophisticated methods are needed.
